File: packages/node/iii-example/src/stream.ts
import { iii } from './iii'
import type { Todo } from './types'

export const streams = {
  get: async <T = unknown>(
    stream_name: string,
    group_id: string,
    item_id: string,
  ): Promise<T | null> => {
    return iii.call('stream::get', { stream_name, group_id, item_id })
  },
  set: async <T = unknown>(
    stream_name: string,
    group_id: string,
    item_id: string,
    data: T,
  ): Promise<T> => {
    return iii.call('stream::set', { stream_name, group_id, item_id, data })
  },
  delete: async (stream_name: string, group_id: string, item_id: string): Promise<void> => {
    return iii.call('stream::delete', { stream_name, group_id, item_id })
  },
  list: async <T = unknown>(stream_name: string, group_id: string): Promise<T[]> => {
    return iii.call('stream::list', { stream_name, group_id })
  },
  listGroups: async (stream_name: string): Promise<string[]> => {
    return iii.call('stream::list_groups', { stream_name })
  },
}

let todoState: Todo[] = []

iii.createStream('todo', {
  get: async input => todoState.find(todo => todo.id === input.item_id),
  set: async input => {
    const existingTodo = todoState.find(todo => todo.id === input.item_id)

    if (existingTodo) {
      const newTodo = { ...existingTodo, ...input.data }
      todoState = todoState.map(todo => (todo.id === input.item_id ? newTodo : todo))
      return { old_value: existingTodo, new_value: newTodo }
    }

    const newTodo: Todo = {
      id: input.item_id,
      groupId: input.group_id,
      description: input.data.description,
      dueDate: input.data.dueDate,
      completedAt: null,
      createdAt: input.data?.createdAt ?? new Date().toISOString(),
    }

    todoState.push(newTodo)

    return { old_value: undefined, new_value: newTodo }
  },
  delete: async input => {
    const old_value = todoState.find(todo => todo.id === input.item_id)
    todoState = todoState.filter(todo => todo.id !== input.item_id)
    return { old_value }
  },
  list: async input => todoState.filter(todo => todo.groupId === input.group_id),
  listGroups: async () => [...new Set(todoState.map(todo => todo.groupId))],
  update: async () => {
    throw new Error('Not implemented')
  },
})

File: packages/python/iii-example/src/stream.py
from __future__ import annotations

from typing import Any

from iii import (
    IStream,
    StreamDeleteInput,
    StreamGetInput,
    StreamListGroupsInput,
    StreamListInput,
    StreamSetInput,
    StreamSetResult,
    StreamUpdateInput,
)

from .iii import get_iii
from .models import Todo


class StreamClient:
    async def get(self, stream_name: str, group_id: str, item_id: str) -> Any | None:
        iii = get_iii()
        return await iii.call(
            "stream::get", {"stream_name": stream_name, "group_id": group_id, "item_id": item_id}
        )

    async def set(self, stream_name: str, group_id: str, item_id: str, data: Any) -> Any:
        iii = get_iii()
        return await iii.call(
            "stream::set", {"stream_name": stream_name, "group_id": group_id, "item_id": item_id, "data": data}
        )

    async def delete(self, stream_name: str, group_id: str, item_id: str) -> None:
        iii = get_iii()
        return await iii.call(
            "stream::delete", {"stream_name": stream_name, "group_id": group_id, "item_id": item_id}
        )

    async def get_group(self, stream_name: str, group_id: str) -> list[Any]:
        iii = get_iii()
        return await iii.call(
            "stream::list", {"stream_name": stream_name, "group_id": group_id}
        )

    async def list_groups(self, stream_name: str) -> list[str]:
        iii = get_iii()
        return await iii.call("stream::list_groups", {"stream_name": stream_name})


class TodoStream(IStream[dict[str, Any]]):
    def __init__(self) -> None:
        self._todos: list[Todo] = []

    async def get(self, input: StreamGetInput) -> dict[str, Any] | None:
        for todo in self._todos:
            if todo.id == input.item_id:
                return todo.model_dump()
        return None

    async def set(self, input: StreamSetInput) -> StreamSetResult[dict[str, Any]] | None:
        for i, todo in enumerate(self._todos):
            if todo.id == input.item_id:
                updated = Todo(**{**todo.model_dump(), **input.data})
                self._todos[i] = updated
                return StreamSetResult(old_value=todo.model_dump(), new_value=updated.model_dump())

        new_todo = Todo(
            id=input.item_id,
            group_id=input.group_id,
            description=input.data.get("description", ""),
            due_date=input.data.get("dueDate"),
            completed_at=None,
        )
        self._todos.append(new_todo)
        return StreamSetResult(old_value=None, new_value=new_todo.model_dump())

    async def delete(self, input: StreamDeleteInput) -> None:
        self._todos = [t for t in self._todos if t.id != input.item_id]

    async def list(self, input: StreamListInput) -> list[dict[str, Any]]:
        return [t.model_dump() for t in self._todos if t.group_id == input.group_id]

    async def list_groups(self, input: StreamListGroupsInput) -> list[str]:
        return list({t.group_id for t in self._todos})

    async def update(self, input: StreamUpdateInput) -> StreamSetResult[dict[str, Any]] | None:
        return None


streams = StreamClient()
_streams_registered = False


def register_streams() -> None:
    global _streams_registered
    if _streams_registered:
        return
    get_iii().create_stream("todo", TodoStream())
    _streams_registered = True
